src/connector/src/parser/chunk_builder.rs

```diff
@@ -14,83 +14,201 @@
+            builder: self,
             visible: true, // write visible rows by default
             row_meta: None,
         }
     }
 
-    /// Consumes the builder and returns a [`StreamChunk`].
-    pub fn finish(self) -> StreamChunk {
-        StreamChunk::with_visibility(
-            self.op_builder,
-            self.builders
+    /// Write a heartbeat record to the builder. The builder will decide whether to finish the
+    /// current chunk or not. Currently it ensures that heartbeats are always in separate chunks.
+    pub fn heartbeat(&mut self, meta: MessageMeta<'_>) {
+        if self.current_chunk_len() > 0 {
+            // If there are records in the chunk, finish it first.
+            // If there's an ongoing transaction, `finish_current_chunk` will handle it properly.
+            // Note this
+            self.finish_current_chunk();
+        }
+
+        _ = self
+            .row_writer()
+            .invisible()
+            .with_meta(meta)
+            .do_insert(|_| Ok(Datum::None));
+        self.finish_current_chunk(); // each heartbeat should be a separate chunk
+    }
+
+    /// Finish and build a [`StreamChunk`] from the current pending records in the builder,
+    /// no matter whether the builder is in a transaction or not, `split_txn` or not. The
+    /// built chunk will be appended to the `ready_chunks` and the builder will be reset.
+    pub fn finish_current_chunk(&mut self) {
+        if self.op_builder.is_empty() {
+            return;
+        }
+
+        let (builders, op_builder, vis_builder) =
+            Self::create_builders(&self.column_descs, self.source_ctrl_opts.chunk_size);
+        let chunk = StreamChunk::with_visibility(
+            std::mem::replace(&mut self.op_builder, op_builder),
+            std::mem::replace(&mut self.builders, builders)
                 .into_iter()
                 .map(|builder| builder.finish().into())
                 .collect(),
-            self.vis_builder.finish(),
-        )
+            std::mem::replace(&mut self.vis_builder, vis_builder).finish(),
+        );
+        self.ready_chunks.push(chunk);
+
+        if let Some(ref mut txn) = self.ongoing_txn {
+            tracing::warn!(
+                txn_id = txn.id,
+                len = txn.len,
+                "splitting an ongoing transaction"
+            );
+            txn.len = 0;
+        }
     }
 
-    /// Resets the builder and returns a [`StreamChunk`], while reserving `next_cap` capacity for
-    /// the builders of the next [`StreamChunk`].
-    #[must_use]
-    pub fn take_and_reserve(&mut self, next_cap: usize) -> StreamChunk {
-        let descs = std::mem::take(&mut self.descs); // we don't use `descs` in `finish`
-        let builder = std::mem::replace(self, Self::with_capacity(descs, next_cap));
-        builder.finish()
+    /// Consumes and returns the ready [`StreamChunk`]s.
+    pub fn consume_ready_chunks(&mut self) -> impl ExactSizeIterator<Item = StreamChunk> + '_ {
```

Comment:
> Rewrite `SourceStreamChunkBuilder` to allow creating chunk during writing to `SourceStreamChunkRowWriter`. The trick is to add a `ready_chunks` field.
>
> This is because that it's possible to parse multiple rows from one `SourceMessage`, while the parser API doesn't use async stream to yield rows, instead, it directly write to `SourceStreamChunkRowWriter`.

I understand the motivation to return multiple chunks, but what's reason of separating `finish()` into `finish_current_chunk() + consume_ready_chunks()`? I mean, if you just want multiple chunks, changing `finish()` to return `Vec<StreamChunk>` can achieve that.

Reply:
Because sometimes we need to call `consume_ready_chunks` separately even we (the user of `SourceStreamChunkBuilder `) don't call `finish_current_chunk` explicitly before it. For example 1) in the arm `ParseResult::Rows` of `match parser.parse_one_with_txn()`; 2) after `chunk_builder.commit_transaction(id)`.